src/clib/pio_darray.cpp

```diff
@@ -3015,17 +3008,28 @@ static int PIOc_read_darray_adios(file_desc_t *file, int fndims, io_desc_t *iode
             }
 
             /* Search for the start and end index of the decomposition map inside one block */
-            for (size_t pos = 0; pos < block_size; pos++)
+            if (len_decomp > 0)
```

Comment:
What if instead of handling decomps with len == 0 here (remove if (len_decomp > 0) checks in this function) let match_decomp_full() return true if len_decomp == 0 & drop the assert for (len_decomp > 0) in the function?
Can the underflow issue with " (pos + len_decomp - 1) < block_size " be handled with type casting (isn't that what the end_idx_in_target_block was using in the old code)?

Reply:
Or maybe,  " (pos + len_decomp) < block_size + 1" etc

Reply:
We should also discuss next week on how we can speed up this matching (at least reading the code here it looks really expensive)? The simplest way to cut the cost for example is to avoid (that can be turned back on via a configure opt) this check for components were we know the read and write decomps are the same.

Reply:
> Or maybe, " (pos + len_decomp) < block_size + 1" etc

Yes, we can use this check to avoid underflow and no type casting is needed.

Reply:
> We should also discuss next week on how we can speed up this matching (at least reading the code here it looks really expensive)? The simplest way to cut the cost for example is to avoid (that can be turned back on via a configure opt) this check for components were we know the read and write decomps are the same.

I already have a plan to further speed up this matching (and no need to guess target blocks) in another PR after this PR is merged. When the decomposition data are not saved to ADIOS BP files, we already can calculate the expected location of the decomp map inside the target block. We only call match_decomp_full function once to confirm. If the matching fails, we know the read and write decomps are different and an error message will be generated.

Reply:
> What if instead of handling decomps with len == 0 here (remove if (len_decomp > 0) checks in this function) let match_decomp_full() return true if len_decomp == 0 & drop the assert for (len_decomp > 0) in the function? Can the underflow issue with " (pos + len_decomp - 1) < block_size " be handled with type casting (isn't that what the end_idx_in_target_block was using in the old code)?

Thanks for your suggestions. I will update this PR to simplify the code changes.